### PR TITLE
Add PlasmaCoilSetMaxDistance objective

### PR DESCRIPTION
**Pull request by ryanwu4**
Sometimes, the coils get too far away from the plasma, and I often need an objective function which constrains the distance between plasma and coil from above. This adds that objective by generalizing the PlasmaCoilSetDistance objective to find the maximum (over the coil) minimum (over the plasma) distance between coil and plasma. To do this, I also added functionality to the `softmax` function to allow it to perform maximization over a given axis rather than flattening the array.

### CODE REVIEW

File: tests/test_objective_funs.py [hunk 1]
```diff
@@ -1249,6 +1250,148 @@ def test(
             coils_fixed=True,
         )
 
+    @pytest.mark.unit
+    def test_plasma_coil_max_distance(self):
```
Comment (dpanici) [inline tests/test_objective_funs.py:1254]: in these cases, is the max also equal to the min?

Might be worth to try shifting one of these slightly outward so that there is an asymmetry introduced in the problem, so that you can check the max stuff more explicitly

**Comment by ryanwu4** [comment]
Not sure why codecov is freaking out over these lines which are actually covered in the tests ...

**Review by dpanici — DISMISSED** [review]

**Comment by ryanwu4** [comment]
Change this to remove Max option, keep Bounds and precompute both max and min so that the bounds objective works the "intuitive" way

File: desc/objectives/_coils.py [hunk 2]
```diff
@@ -1095,12 +1110,135 @@ def body(k):
...
+                return max
+            if self._ismin:
+                return min
+            return jnp.array([min, max])
 
         k = jnp.arange(self.dim_f)
-        min_dist_per_coil = vmap_chunked(body, chunk_size=self._dist_chunk_size)(k)
-        return min_dist_per_coil
+        extreme_dist_per_coil = vmap_chunked(body, chunk_size=self._dist_chunk_size)(k)
+        return extreme_dist_per_coil
```
Comment (YigitElma) [inline desc/objectives/_coils.py:1131]: For ``bound`` case, is this going to return a 2 by N array? I don't know how that would work out for optimization. Can you concatenate this into a single 1D array of size 2*N? Also, you should change the `dim_f` part to account for the special ``bound`` case, and most probably the target to be an array like [min, max, min, max, ...]

**Review by YigitElma — CHANGES_REQUESTED** [review]

File: desc/objectives/_coils.py (hunk 2 above)
Comment (YigitElma) [reply desc/objectives/_coils.py:1131]: The ``target`` can also be in [min, min, min,...., max, max, max ...] order. Whichever you choose

**Review by YigitElma — APPROVED** [review]

File: desc/objectives/_coils.py [hunk 3]
```diff
@@ -931,16 +940,23 @@ def __init__(
...
         if target is None and bounds is None:
             bounds = (1, np.inf)
+        errorif(
+            mode not in ["bound", "max", "min"],
+            ValueError,
+            "mode must be 'bound', 'max', or 'min'",
+        )
         self._eq = eq
         self._coil = coil
+        self._ismax = mode == "max"
```
Comment (f0uriest) [inline desc/objectives/_coils.py:958]: Why not just store mode itself instead of two bools?

File: desc/objectives/_coils.py [hunk 4]
```diff
@@ -931,16 +940,23 @@ def __init__(
...
         eq_fixed=False,
         coils_fixed=False,
-        name="plasma-coil minimum distance",
+        name="plasma-coil distance",
         jac_chunk_size=None,
         use_softmin=False,
         softmin_alpha=1.0,
         dist_chunk_size=None,
     ):
         if target is None and bounds is None:
```
Comment (f0uriest) [inline desc/objectives/_coils.py:949]: Do we want the default for this to be different from PlasmaCoilsetMinDistance?

File: desc/objectives/_coils.py [hunk 5]
```diff
@@ -850,11 +850,15 @@ def body(k):
...
 
 
-class PlasmaCoilSetMinDistance(_Objective):
-    """Target the minimum distance between the plasma and coilset.
+class PlasmaCoilSetDistanceBound(_Objective):
+    """Target the distance between the plasma and coilset.
 
-    Will yield one value per coil in the coilset, which is the minimum distance from
-    that coil to the plasma boundary surface.
+    Will yield one or two values per coil in the coilset, depending on the mode
```
Comment (f0uriest) [inline desc/objectives/_coils.py:856]: Should mention somewhere here that the minima is the first N values in the array, then the maxima, it's not interleaved

**Comment by YigitElma** [comment]
Can we also consider solving #1733 here?

**Review by dpanici — APPROVED** [review]

**Review by f0uriest — APPROVED** [review]
